### PR TITLE
Navigation: Implement active state for items in the Sidemenu

### LINKED ISSUE #39021: Sidemenu: Implement correct hover/focus/current state
**Why is this component needed**:

**Situation:** 
- Right now hover and selected states are visually similar – they both employ gradient borders. We aim to visually distinguish these two states as much as possible.
- Grafana currently lacks a clear indicator of which navbar item is selected, and where in the navigation our users are. This makes wayfinding problematic.
- Focus state is clunky. The left border of the blue focus ring spills over the screen. It also lacks prominence. Tabbing is unpredictable, which is why focus states should be more visually reinforced.

**Goals:** 

- Design a hover state that is visually different than the selected state.
- Design a more prominent focused state.
- Make it clear which section of the navbar is selected so that users know where they are in the navigation by designing a selected state.
- Solutions will be consistent with our design system and patterns that Grafana is currently employing, and it will be in line with best practices of UI design.

**Solutions:**

- All states have been checked for accessibility.
- All states employ a double indicator of state change: e.g. overlay + icon color; icon color + border.
- Hover state uses a dark overlay, and the default icon color changes from secondary to primary. It has low visual emphasis as the cursors assists users in tracking where they are.
- Focus state adds a focus ring on top of the hover state. Focus ring is inside the element, it does not spill outside of the navbar. It has medium visual emphasis as tabbing is unpredictable and there is no cursor to assist orientation.
- Selected state utilises our primary icon color and a gradient border on the left side of the element container. This
 is consistent with how we indicate selected tabs in our UI. It has medium emphasis that allows users to quickly understand where they are, but does not take away from the main content.

Screenshot 2021-09-08 at 14 50 26
___

This will be used in the navigation sidebar for both the light and the dark theme.
For https://github.com/grafana/grafana/issues/37961

as per the designs from @VesnaDean from [thisFigma file](https://www.figma.com/file/m0q3sBVILBNGZO1QolsMn6/Navbar-explorations?node-id=0%3A1)

### PR DESCRIPTION
**Pull request by ashharrison90**
**What this PR does / why we need it**:

### tweaks hover state to match new designs:
 

**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/grafana/issues/39021

**Special notes for your reviewer**:
- there is one thing that doesn't work so well - there is nothing "active" in the sidemenu when viewing a dashboard. that's because none of the routes in the `dashboard` link match any particular dashboard route - this probably needs some further discussion to work out the best solution.

### CODE REVIEW

**Comment by VesnaDean** [comment]
Looks on point to me. My only question is about the light theme: Does the border color change to light theme's primary for the focus ring?

**Comment by ashharrison90** [comment]
> Looks on point to me. My only question is about the light theme: Does the border color change to light theme's primary for the focus ring?

yup. it uses the primary theme variable so all that is handled automatically 👍 

image

File: public/app/core/components/sidemenu/TopSection.tsx [hunk 1]
```diff
@@ -1,13 +1,18 @@
...
 import { locationService } from '@grafana/runtime';
 import { Icon, IconName, styleMixins, useTheme2 } from '@grafana/ui';
 import config from '../../config';
+import { linkIsActive } from './utils';
 import SideMenuItem from './SideMenuItem';
 
 const TopSection = () => {
+  const location = useLocation();
+  const query = new URLSearchParams(location.search);
+  const isSearchActive = query.get('search') === 'open';
```
Comment (hugohaggmark) [inline public/app/core/components/sidemenu/TopSection.tsx:15]: Might want to consider wrapping this logic up into a custom hook or utility function?

**Review by hugohaggmark — APPROVED** [review]
LGTM and works as expected 🎉   I think the selected state indicator looks a bit too thick (It's according to the design, so this is my preference).

image

Any ideas on how to handle the selected state for dashboards or are you opening a separate PR for that?

**Comment by hugohaggmark** [comment]
Hmm, I found another thing. The selected state for some settings pages for entities further down doesn't work.
http://localhost:3000/datasources/edit/8HhfGHEGk (edit a data source)
http://localhost:3000/alerting/notification/2/edit (edit a notification channel)

**Comment by ashharrison90** [comment]
thanks @hugohaggmark 🙌  i'll work on fixing those issues.

for the selected state for dashboards, my plan was to discuss it with everyone else and see what the best approach would be.

we could just put some custom logic in that says anything under the `/d/` route is a dashboard, or we could move dashboards under the `/dashboards/` route for consistency and implement a redirect from `/d/` to `/dashboards/` to maintain backwards compatibility. i'd like to discuss it anyway to see what everyone's thoughts are 👍

**Comment by hugohaggmark** [comment]
> thanks @hugohaggmark 🙌 i'll work on fixing those issues.
> 
> for the selected state for dashboards, my plan was to discuss it with everyone else and see what the best approach would be.
> 
> we could just put some custom logic in that says anything under the `/d/` route is a dashboard, or we could move dashboards under the `/dashboards/` route for consistency and implement a redirect from `/d/` to `/dashboards/` to maintain backwards compatibility. i'd like to discuss it anyway to see what everyone's thoughts are 👍

I think we should start with `/d/` and discuss the redirect solution later.

**Review by hugohaggmark — COMMENTED** [review]
Great work 🎉 

I found a new minor issue when choosing create folder 😄 
image and still no indicator for notification channels.

image

**Comment by ashharrison90** [comment]
thanks again @hugohaggmark, this is a really great review 🙌 

i _think_ everything's working now, but the code feels very brittle. might be worth having a mini sync on how we can refactor routes such that this can be more generic. wdyt? 🤔

**Review by axelavargas — APPROVED** [review]
🙌🏾 great work on this PR @ashharrison90 (The side menu is looking way better with all of the UX improvements from @VesnaDean) 👏🏾 , I found a minor issue while testing, but besides that everything works as expected.

File: public/app/core/components/sidemenu/utils.ts [hunk 2]
```diff
@@ -1,8 +1,42 @@
...
+    return true;
+  } else if (link.url === '/alerting/list' && pathname.startsWith('/alerting/notification/')) {
+    // alert channel match
+    // TODO refactor routes such that we don't need this custom logic
+    return true;
+  } else if (link.url === '/' && pathname.startsWith('/d/')) {
+    // dashboard match
+    // TODO refactor routes such that we don't need this custom logic
+    return true;
+  } else if (link.children?.some((childLink) => isLinkActive(pathname, childLink))) {
```
[the remark this answers is not in the document]
Comment (ashharrison90) [reply public/app/core/components/sidemenu/utils.ts:32]: thanks @axelavargas 🙌  working on a fix now 😄

**Comment by ashharrison90** [comment]
@axelavargas thanks again! 🥳 

i've stripped the query params out from the link and it should match properly now.
let me know what you think! 😄

**Review by axelavargas — APPROVED** [review]
It is working correctly with the plugin's URLs 🚀, great job!

**Pull request merged**